### ISSUE TITLE
Crash calling 2D space intersect_shape with 3D query parameters

### ISSUE THREAD

**Issue by Jackain** [godotengine/godot#17439]
**Godot version:**
Godot 3.0.2 Stable

**OS/device including version:**
Win 7 x64

**Issue description:**
When the project run the intersect_shape(  function, Godot crash with this error : 

CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://git
hub.com/godotengine/godot/issues
[0] <couldn't map PC to fn name>
[1] <couldn't map PC to fn name>
[2] <couldn't map PC to fn name>
[3] <couldn't map PC to fn name>
[4] <couldn't map PC to fn name>
[5] <couldn't map PC to fn name>
[6] <couldn't map PC to fn name>
[7] <couldn't map PC to fn name>
[8] <couldn't map PC to fn name>
[9] <couldn't map PC to fn name>
[10] <couldn't map PC to fn name>
[11] <couldn't map PC to fn name>
[12] <couldn't map PC to fn name>
[13] <couldn't map PC to fn name>
[14] <couldn't map PC to fn name>
[15] <couldn't map PC to fn name>
[16] <couldn't map PC to fn name>
[17] <couldn't map PC to fn name>
[18] <couldn't map PC to fn name>
[19] <couldn't map PC to fn name>
[20] <couldn't map PC to fn name>
[21] <couldn't map PC to fn name>
[22] <couldn't map PC to fn name>
[23] BaseThreadInitThunk
-- END OF BACKTRACE --
Nothing read: Invalid argument
ERROR: StreamPeerTCPWinsock::read: Server disconnected!

   At: drivers\windows\stream_peer_tcp_winsock.cpp:203

**Steps to reproduce:**
Just run the project and play

**Minimal reproduction project:**
[Bug raycast.zip](https://github.com/godotengine/godot/files/1800417/Bug.raycast.zip)

Thank you !

**Comment by ShyRed** [comment]
Holy cow, that was a nasty one to figure out :joy:

You are working in **2D** space which means you need to use the **2D** objects. Replace `PhysicsShapeQueryParameters` with `Physics2DShapeQueryParameters` and `set_collision_mask` with `set_collision_layer` and it works fine.

The real issue here is that Godot does not notice that the type is wrong. The crash is because down the line it tries to convert PhysicsShapeQueryParameters into Physics2DShapeQueryParameters which fails silently and results in NULL being passed to the Physics2DDirectSpaceState::_intersect_shape method. As the method does not check for NULL it crashes when trying to access the parameter.

I do think that this is a general problem. This most likely will be fixed once GDScript supports typing ( see #10630 )

Working script:
```GDScript
func segment_cast(begin_pos, end_pos):
	var space_state = ray.get_world_2d().get_direct_space_state()
	var segment = SegmentShape2D.new()
	segment.set_a(begin_pos)
	segment.set_b(end_pos)
	var query = Physics2DShapeQueryParameters.new()
	query.set_exclude([self])
	query.set_collision_layer(2)
	query.set_shape(segment)
	var hits = space_state.intersect_shape(query, 12)
```

**Comment by akien-mga** [comment]
> I do think that this is a general problem. This most likely will be fixed once GDScript supports typing ( see #10630 )

Well in this case it's just a missing null check though. GDScript typing will be optional, so it won't solve type mismatches in the case where people don't use the optional typing.

**Comment by vnen** [comment]
@reduz the issue is not in GDScript, but in the physics engine (missing an argument check).
